File: leetcode/0472-concatenated-words/solution.py
from typing import List


class Solution:
    def findAllConcatenatedWordsInADict(self, words: List[str]) -> List[str]:
        word_set = set(words)
        memo = {}

        def canBuild(word: str) -> bool:
            # can `word` be fully assembled from >= 2 other words in word_set?
            if word in memo:
                return memo[word]
            memo[word] = False  # guard against reuse while computing (no cycles expected, but safe)
            n = len(word)
            dp = [False] * (n + 1)
            dp[0] = True
            for i in range(1, n + 1):
                for j in range(0, i):
                    if not dp[j]:
                        continue
                    piece = word[j:i]
                    # piece must be a real word other than `word` itself when it spans the whole thing
                    if piece == word:
                        continue
                    if piece in word_set:
                        dp[i] = True
                        break
            memo[word] = dp[n]
            return dp[n]

        result = []
        for w in words:
            if len(w) == 0:
                continue
            if canBuild(w):
                result.append(w)
        return result

File: leetcode/0472-concatenated-words/solution.cpp
#include <vector>
#include <string>
#include <unordered_set>
using namespace std;

class Solution {
public:
    vector<string> findAllConcatenatedWordsInADict(vector<string>& words) {
        unordered_set<string> wordSet(words.begin(), words.end());

        vector<string> result;
        for (const string& w : words) {
            if (w.empty()) continue;
            if (canBuild(w, wordSet)) result.push_back(w);
        }
        return result;
    }

private:
    bool canBuild(const string& word, unordered_set<string>& wordSet) {
        int n = (int)word.size();
        vector<bool> dp(n + 1, false);
        dp[0] = true;
        for (int i = 1; i <= n; i++) {
            for (int j = 0; j < i; j++) {
                if (!dp[j]) continue;
                string piece = word.substr(j, i - j);
                if (piece == word) continue;
                if (wordSet.find(piece) != wordSet.end()) {
                    dp[i] = true;
                    break;
                }
            }
        }
        return dp[n];
    }
};
